# Configure --------------------------------------------------------------------

configfile: "config.yaml"
STUDIES, = glob_wildcards(config["ss_file_pattern"])

# Snakemake rules --------------------------------------------------------------

rule all:
    input: 
        expand("output/bp_{step}_chr_{chromosome}.{ss_file}.output.tsv",
               ss_file=STUDIES, step=config["steps"], chromosome=config["chromosomes"])


rule get_vcf_files:
    # these are for build 37 
    output:
        "gnomad.genomes.r2.0.2.sites.chr{chromosome}.vcf.bgz"
    params:
        location=config["remote_vcf_location"],
    shell:
        "wget {params.location}/gnomad.genomes.r2.0.2.sites.chr{wildcards.chromosome}.vcf.bgz"


rule get_tbi_files:
    # these are for build 37 -if not available, we can use tabix to index
    output:
        "gnomad.genomes.r2.0.2.sites.chr{chromosome}.vcf.bgz.tbi"
    params:
        location=config["remote_vcf_location"],
    shell:
        "wget {params.location}/gnomad.genomes.r2.0.2.sites.chr{wildcards.chromosome}.vcf.bgz.tbi"


rule split_by_chrom:
    # split the sumstat files by chromosome
    input:
        expand("ss_files/{ss_file}.tsv", ss_file=STUDIES)
    output:
        expand("chr_{chromosome}.{{ss_file}}.tsv", chromosome=config["chromosomes"])
    params:
        chromosome=config["chromosomes"]
    shell:
        "python split_by_chromosome.py "
        "-f {input} "
        "-chr {params.chromosome}"

rule split_by_bp:
    # split chromosome files into 16ths
    input:
        "chr_{chromosome}.{ss_file}.tsv"
    output:
        expand("bp_{step}.chr_{{chromosome}}.{{ss_file}}.tsv", step=config["steps"]),
        expand("run_bp_{step}.chr_{{chromosome}}.{{ss_file}}.sh", step=config["steps"])
    shell:
        "python split_by_bp.py "
        "-f {input} "
        "-chr {wildcards.chromosome}"

rule run_harmonisation_per_split:
    # run sumstat_harmoniser.py for each split
    input:
        "gnomad.genomes.r2.0.2.sites.chr{chromosome}.vcf.bgz",
        "gnomad.genomes.r2.0.2.sites.chr{chromosome}.vcf.bgz.tbi",
        "bp_{step}.chr_{chromosome}.{ss_file}.tsv",
        command="run_bp_{step}.chr_{chromosome}.{ss_file}.sh"  
    output:
        "output/bp_{step}.chr_{chromosome}.{ss_file}.output.tsv"
    shell:
        "source {input.command}"

